File: smells/unrealistic/surreal/surreal.js
/* Smell: Surreal tests
 */

// Subject under test
var _ = require('lodash')
function weighClothes (clothes) {
  return Math.round(_.sumBy(clothes, function (item) {
    return item.weight()
  }))
}

// Test
var td = require('testdouble')
module.exports = {
  beforeEach: function () {
    td.replace(factors, 'size')
    td.when(factors.size('S')).thenReturn(9)
  },
  addsWeights: function () {
    var smallWetSock = new Clothing('S', 'sock', 'wet')
    var largeDryJacket = new Clothing('L', 'jacket', 'dry')
    td.when(td.replace(largeDryJacket, 'weight')()).thenReturn(8)
    var xlSoakedPants = td.object(['weight'])
    td.when(xlSoakedPants.weight()).thenReturn(15)

    var result = weighClothes([smallWetSock, largeDryJacket, xlSoakedPants])

    assert.equal(result, 26)
  },
  afterEach: function () {
    td.reset()
  }
}

// Fake production implementations to simplify example test of subject
function Clothing (size, type, wetness) {
  this.size = size
  this.type = type
  this.wetness = wetness
}
Clothing.prototype.weight = function () {
  return 1 *
    factors.size(this.size) *
    factors.type(this.type) *
    factors.wetness(this.wetness)
}

var factors = {
  size: function (size) {
    switch (size) {
      case 'S':
        return 0.75
      case 'M':
        return 1
      case 'L':
        return 1.25
      case 'XL':
        return 1.5
      default:
        return 1
    }
  },
  type: function (type) {
    switch (type) {
      case 'sock':
        return 0.2
      case 'shirt':
        return 1
      case 'pants':
        return 2
      case 'jacket':
        return 3
      default:
        return 1
    }
  },
  wetness: function (wetness) {
    switch (wetness) {
      case 'dry':
        return 1
      case 'moist':
        return 1.1
      case 'wet':
        return 1.6
      case 'soaked':
        return 2.5
      default:
        return 1
    }
  }
}

File: smells/unrealistic/surreal/surreal.rb
# Subject under test
def weigh_clothes(clothes)
  clothes.map { |item| item.weight }.reduce(:+).round
end

# Test
class Surreal < SmellTest
  def setup
    stub(FACTORS, :size, 9, ["S"])
    super
  end

  def test_adds_weights
    small_wet_sock = Clothing.new("S", "sock", "wet")
    large_dry_jacket = Clothing.new("L", "jacket", "dry")
    stub(large_dry_jacket, :weight, 8)
    xl_soaked_pants = OpenStruct.new(weight: 15)

    result = weigh_clothes([small_wet_sock, large_dry_jacket, xl_soaked_pants])

    assert_equal 26, result
  end
end

# Fake production implementations to simplify example test of subject
class Clothing
  def initialize(size, type, wetness)
    @size = size
    @type = type
    @wetness = wetness
  end

  def weight
    return 1 *
      FACTORS.size(@size) *
      FACTORS.type(@type) *
      FACTORS.wetness(@wetness)
  end
end

class Factors
  def size(size)
    case size
    when "S" then 0.75
    when "M" then 1
    when "L" then 1.25
    when "XL" then 1.5
    else 1
    end
  end

  def type(type)
    case type
    when "sock" then 0.2
    when "shirt" then 1
    when "pants" then 2
    when "jacket" then 3
    else 1
    end
  end

  def wetness(wetness)
    case wetness
    when "dry" then 1
    when "moist" then 1.1
    when "wet" then 1.6
    when "soaked" then 2.5
    else 1
    end
  end
end
FACTORS = Factors.new
